MigratedData killing a server migrates data to the other node, migrated data has new uuid, near cache discards data with old uuid

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/NodeJS-8/304/console
```
08:50:28   MigratedData
08:50:36 [DefaultLogger] INFO at ClusterService: Members received.
08:50:36 [ Member {
08:50:36     address: Address { host: 'localhost', port: 5701, type: 0 },
08:50:36     uuid: 'a31c646f-7201-48b1-a342-751352f5401b',
08:50:36     isLiteMember: false,
08:50:36     attributes: {} },
08:50:36   Member {
08:50:36     address: Address { host: 'localhost', port: 5702, type: 0 },
08:50:36     uuid: '5d4004e2-f8f4-4cae-bc9e-df44ae153d26',
08:50:36     isLiteMember: false,
08:50:36     attributes: {} } ]
08:50:36 [DefaultLogger] INFO at HazelcastClient: Client started
08:50:37 [DefaultLogger] INFO at ClusterService: Member[ uuid: 5d4004e2-f8f4-4cae-bc9e-df44ae153d26, address: localhost:5702] removed from cluster
08:50:37 [DefaultLogger] WARN at ClusterService: Connection closed to localhost:5702
08:50:39     1) killing a server migrates data to the other node, migrated data has new uuid, near cache discards data with old uuid

...

08:51:46   904 passing (10m)
08:51:46   53 pending
08:51:46   1 failing
08:51:46 
08:51:46   1) MigratedData killing a server migrates data to the other node, migrated data has new uuid, near cache discards data with old uuid:
08:51:46 
08:51:46       AssertionError: expected 0 to equal 1
08:51:46       + expected - actual
08:51:46 
08:51:46       -0
08:51:46       +1
08:51:46       
08:51:46       at test/nearcache/MigratedDataTest.js:104:39
08:51:46       at tryCatcher (node_modules/bluebird/js/release/util.js:16:23)
08:51:46       at Promise._settlePromiseFromHandler (node_modules/bluebird/js/release/promise.js:512:31)
08:51:46       at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:569:18)
08:51:46       at Promise._settlePromise0 (node_modules/bluebird/js/release/promise.js:614:10)
08:51:46       at Promise._settlePromises (node_modules/bluebird/js/release/promise.js:693:18)
08:51:46       at Async._drainQueue (node_modules/bluebird/js/release/async.js:133:16)
08:51:46       at Async._drainQueues (node_modules/bluebird/js/release/async.js:143:10)
08:51:46       at Immediate.Async.drainQueues (node_modules/bluebird/js/release/async.js:17:14)
```

--- a/src/nearcache/MetadataFetcher.ts
+++ b/src/nearcache/MetadataFetcher.ts
@@ -37,11 +37,13 @@ export class MetadataFetcher {
     }
 
     initHandler(handler: RepairingHandler): Promise<void> {
-        const scanPromise = this.scanMembers([handler.getName()])[0];
-        return scanPromise.then((response: ClientMessage) => {
-            const metadata = MapFetchNearCacheInvalidationMetadataCodec.decodeResponse(response);
-            handler.initUuid(metadata.partitionUuidList);
-            handler.initSequence(metadata.namePartitionSequenceList[0]);
+        const scanPromises = this.scanMembers([handler.getName()]);
+        return Promise.all(scanPromises).then((responses: ClientMessage[]) => {
+            responses.forEach((response) => {
+                const metadata = MapFetchNearCacheInvalidationMetadataCodec.decodeResponse(response);
+                handler.initUuid(metadata.partitionUuidList);
+                handler.initSequence(metadata.namePartitionSequenceList[0]);
+            });
         });
     }
 

--- a/src/nearcache/NearCache.ts
+++ b/src/nearcache/NearCache.ts
@@ -25,6 +25,7 @@ import {shuffleArray} from '../Util';
 import * as AlwaysFreshStaleReadDetectorImpl from './AlwaysFreshStaleReadDetectorImpl';
 import {DataRecord} from './DataRecord';
 import {StaleReadDetector} from './StaleReadDetector';
+import * as Promise from 'bluebird';
 
 export interface NearCacheStatistics {
     evictedCount: number;
@@ -37,7 +38,7 @@ export interface NearCacheStatistics {
 export interface NearCache {
     put(key: Data, value: any): void;
 
-    get(key: Data): Data | any;
+    get(key: Data): Promise<Data | any>;
 
     invalidate(key: Data): void;
 
@@ -52,6 +53,8 @@ export interface NearCache {
     tryReserveForUpdate(key: Data): Long;
 
     tryPublishReserved(key: Data, value: any, reservationId: Long): any;
+
+    setReady(): void;
 }
 
 export class NearCacheImpl implements NearCache {
@@ -76,6 +79,7 @@ export class NearCacheImpl implements NearCache {
     private missCount: number = 0;
     private hitCount: number = 0;
     private compareFunc: (x: DataRecord, y: DataRecord) => number;
+    private ready: Promise.Resolver<void>;
 
     constructor(nearCacheConfig: NearCacheConfig, serializationService: SerializationService) {
         this.serializationService = serializationService;
@@ -100,6 +104,11 @@ export class NearCacheImpl implements NearCache {
 
         this.evictionCandidatePool = [];
         this.internalStore = new DataKeyedHashMap<DataRecord>();
+        this.ready = Promise.defer();
+    }
+
+    setReady(): void {
+        this.ready.resolve();
     }
 
     nextReservationId(): Long {
@@ -173,30 +182,32 @@ export class NearCacheImpl implements NearCache {
      * @param key
      * @returns the value if present in near cache, 'undefined' if not
      */
-    get(key: Data): Data | any {
-        const dr = this.internalStore.get(key);
-        if (dr === undefined) {
-            this.missCount++;
-            return undefined;
-        }
-        if (this.staleReadDetector.isStaleRead(key, dr)) {
-            this.internalStore.delete(key);
-            this.missCount++;
-            return undefined;
-        }
-        if (dr.isExpired(this.maxIdleSeconds)) {
-            this.expireRecord(key);
-            this.missCount++;
-            return undefined;
-        }
-        dr.setAccessTime();
-        dr.hitRecord();
-        this.hitCount++;
-        if (this.inMemoryFormat === InMemoryFormat.BINARY) {
-            return this.serializationService.toObject(dr.value);
-        } else {
-            return dr.value;
-        }
+    get(key: Data): Promise<Data | any> {
+        return this.ready.promise.then(() => {
+            const dr = this.internalStore.get(key);
+            if (dr === undefined) {
+                this.missCount++;
+                return undefined;
+            }
+            if (this.staleReadDetector.isStaleRead(key, dr)) {
+                this.internalStore.delete(key);
+                this.missCount++;
+                return undefined;
+            }
+            if (dr.isExpired(this.maxIdleSeconds)) {
+                this.expireRecord(key);
+                this.missCount++;
+                return undefined;
+            }
+            dr.setAccessTime();
+            dr.hitRecord();
+            this.hitCount++;
+            if (this.inMemoryFormat === InMemoryFormat.BINARY) {
+                return this.serializationService.toObject(dr.value);
+            } else {
+                return dr.value;
+            }
+        });
     }
 
     invalidate(key: Data): void {

--- a/src/nearcache/RepairingHandler.ts
+++ b/src/nearcache/RepairingHandler.ts
@@ -101,6 +101,7 @@ export class RepairingHandler {
         if (currentUuid != null && currentUuid.equals(newuuid)) {
             return;
         }
+
         metadata.setUuid(newuuid);
         metadata.reset();
     }

--- a/src/nearcache/RepairingTask.ts
+++ b/src/nearcache/RepairingTask.ts
@@ -21,6 +21,7 @@ import {LoggingService} from '../logging/LoggingService';
 import {MetadataFetcher} from './MetadataFetcher';
 import {NearCache} from './NearCache';
 import {RepairingHandler} from './RepairingHandler';
+import * as Promise from 'bluebird';
 
 const PROPERTY_MAX_RECONCILIATION_INTERVAL_SECONDS = 'hazelcast.invalidation.reconciliation.interval.seconds';
 const PROPERTY_MIN_RECONCILIATION_INTERVAL_SECONDS = 'hazelcast.invalidation.min.reconciliation.interval.seconds';
@@ -52,18 +53,20 @@ export class RepairingTask {
         this.partitionCount = this.client.getPartitionService().getPartitionCount();
     }
 
-    registerAndGetHandler(objectName: string, nearCache: NearCache): RepairingHandler {
+    registerAndGetHandler(objectName: string, nearCache: NearCache): Promise<RepairingHandler> {
         let handler = this.handlers.get(objectName);
         if (handler !== undefined) {
-            return handler;
+            return Promise.resolve(handler);
         }
+
         handler = new RepairingHandler(objectName, this.client.getPartitionService(), nearCache, this.localUuid);
-        this.metadataFetcher.initHandler(handler);
-        this.handlers.set(objectName, handler);
-        if (this.antientropyTaskHandle === undefined) {
-            this.start();
-        }
-        return handler;
+        return this.metadataFetcher.initHandler(handler).then(() => {
+            this.handlers.set(objectName, handler);
+            if (this.antientropyTaskHandle === undefined) {
+                this.start();
+            }
+            return handler;
+        });
     }
 
     deregisterHandler(objectName: string): void {

--- a/src/proxy/NearCachedMapProxy.ts
+++ b/src/proxy/NearCachedMapProxy.ts
@@ -43,7 +43,10 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
         if (this.nearCache.isInvalidatedOnChange()) {
             this.addNearCacheInvalidationListener().then((id: string) => {
                 this.invalidationListenerId = id;
+                this.nearCache.setReady();
             });
+        } else {
+            this.nearCache.setReady();
         }
     }
 
@@ -57,12 +60,13 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
     }
 
     protected containsKeyInternal(keyData: Data): Promise<boolean> {
-        const cachedValue = this.nearCache.get(keyData);
-        if (cachedValue !== undefined) {
-            return Promise.resolve(cachedValue != null);
-        } else {
-            return super.containsKeyInternal(keyData);
-        }
+        return this.nearCache.get(keyData).then((cachedValue) => {
+            if (cachedValue !== undefined) {
+                return Promise.resolve(cachedValue != null);
+            } else {
+                return super.containsKeyInternal(keyData);
+            }
+        });
     }
 
     protected deleteInternal(keyData: Data): Promise<void> {
@@ -110,18 +114,20 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
     }
 
     protected getInternal(keyData: Data): Promise<V> {
-        const cachedValue = this.nearCache.get(keyData);
-        if (cachedValue !== undefined) {
-            return Promise.resolve(cachedValue);
-        } else {
-            const reservation = this.nearCache.tryReserveForUpdate(keyData);
-            return super.getInternal(keyData).then((val: V) => {
-                this.nearCache.tryPublishReserved(keyData, val, reservation);
-                return val;
-            }).catch((err: any) => {
-                throw err;
-            });
-        }
+        return this.nearCache.get(keyData).then((cachedValue) => {
+            if (cachedValue !== undefined) {
+                return Promise.resolve(cachedValue);
+            } else {
+                const reservation = this.nearCache.tryReserveForUpdate(keyData);
+                return super.getInternal(keyData).then((val: V) => {
+
+                    this.nearCache.tryPublishReserved(keyData, val, reservation);
+                    return val;
+                }).catch((err: any) => {
+                    throw err;
+                });
+            }
+        });
     }
 
     protected tryRemoveInternal(keyData: Data, timeout: number): Promise<boolean> {
@@ -133,35 +139,39 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
     }
 
     protected getAllInternal(partitionsToKeys: { [id: string]: any }, result: any[] = []): Promise<any[]> {
+        const promises = [];
         try {
             for (const partition in partitionsToKeys) {
                 let partitionArray = partitionsToKeys[partition];
                 for (let i = partitionArray.length - 1; i >= 0; i--) {
                     const key = partitionArray[i];
-                    const cachedResult = this.nearCache.get(key);
-                    if (cachedResult !== undefined) {
-                        result.push([this.toObject(partitionArray[i]), cachedResult]);
-                        partitionArray = partitionArray.splice(i, 1);
-                    }
+                    promises.push(this.nearCache.get(key).then((cachedResult) => {
+                        if (cachedResult !== undefined) {
+                            result.push([this.toObject(partitionArray[i]), cachedResult]);
+                            partitionArray = partitionArray.splice(i, 1);
+                        }
+                    }));
                 }
             }
         } catch (err) {
             return Promise.resolve([]);
         }
-        const reservations: Long[] = [];
-        for (const partition in partitionsToKeys) {
-            const partitionArray = partitionsToKeys[partition];
-            for (const key of partitionArray) {
-                reservations.push(this.nearCache.tryReserveForUpdate(key));
+        return Promise.all(promises).then(() => {
+            const reservations: Long[] = [];
+            for (const partition in partitionsToKeys) {
+                const partitionArray = partitionsToKeys[partition];
+                for (const key of partitionArray) {
+                    reservations.push(this.nearCache.tryReserveForUpdate(key));
+                }
             }
-        }
-        return super.getAllInternal(partitionsToKeys, result).then((serializedEntryArray: Array<[Data, Data]>) => {
-            serializedEntryArray.forEach((serializedEntry: [Data, Data], index: number) => {
-                const key = serializedEntry[0];
-                const value = serializedEntry[1];
-                this.nearCache.tryPublishReserved(key, value, reservations[index]);
+            return super.getAllInternal(partitionsToKeys, result).then((serializedEntryArray: Array<[Data, Data]>) => {
+                serializedEntryArray.forEach((serializedEntry: [Data, Data], index: number) => {
+                    const key = serializedEntry[0];
+                    const value = serializedEntry[1];
+                    this.nearCache.tryPublishReserved(key, value, reservations[index]);
+                });
+                return result;
             });
-            return result;
         });
     }
 
@@ -200,7 +210,9 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
     private addNearCacheInvalidationListener(): Promise<string> {
         const codec = this.createInvalidationListenerCodec(this.name, EntryEventType.INVALIDATION);
         if (this.supportsRepairableNearCache()) {
-            return this.client.getListenerService().registerListener(codec, this.createNearCacheEventHandler());
+            return this.createNearCacheEventHandler().then((handler) => {
+                return this.client.getListenerService().registerListener(codec, handler);
+            });
         } else {
             return this.client.getListenerService().registerListener(codec, this.createPre38NearCacheEventHandler());
         }
@@ -259,20 +271,21 @@ export class NearCachedMapProxy<K, V> extends MapProxy<K, V> {
         };
     }
 
-    private createNearCacheEventHandler() {
+    private createNearCacheEventHandler(): Promise<Function> {
         const repairingTask = this.client.getRepairingTask();
-        const repairingHandler = repairingTask.registerAndGetHandler(this.getName(), this.nearCache);
-        const staleReadDetector = new StaleReadDetectorImpl(repairingHandler, this.client.getPartitionService());
-        this.nearCache.setStaleReadDetector(staleReadDetector);
-        const handle = function (key: Data, sourceUuid: string, partitionUuid: UUID, sequence: Long) {
-            repairingHandler.handle(key, sourceUuid, partitionUuid, sequence);
-        };
-        const handleBatch = function (keys: Data[], sourceUuids: string[], partititonUuids: UUID[], sequences: Long[]) {
-            repairingHandler.handleBatch(keys, sourceUuids, partititonUuids, sequences);
-        };
+        return repairingTask.registerAndGetHandler(this.getName(), this.nearCache).then((repairingHandler) => {
+            const staleReadDetector = new StaleReadDetectorImpl(repairingHandler, this.client.getPartitionService());
+            this.nearCache.setStaleReadDetector(staleReadDetector);
+            const handle = function (key: Data, sourceUuid: string, partitionUuid: UUID, sequence: Long) {
+                repairingHandler.handle(key, sourceUuid, partitionUuid, sequence);
+            };
+            const handleBatch = function (keys: Data[], sourceUuids: string[], partititonUuids: UUID[], sequences: Long[]) {
+                repairingHandler.handleBatch(keys, sourceUuids, partititonUuids, sequences);
+            };
 
-        return function (m: ClientMessage) {
-            MapAddNearCacheInvalidationListenerCodec.handle(m, handle, handleBatch);
-        };
+            return function (m: ClientMessage) {
+                MapAddNearCacheInvalidationListenerCodec.handle(m, handle, handleBatch);
+            };
+        });
     }
 }

--- a/test/nearcache/NearCacheTest.js
+++ b/test/nearcache/NearCacheTest.js
@@ -145,18 +145,22 @@ describe('NearCacheImpl', function () {
 
             beforeEach(function () {
                 nearCache = new NearCacheImpl(testConfig, createSerializationService());
+                nearCache.setReady();
             });
 
 
             it('simple put/get', function () {
                 nearCache.put(ds('key'), 'val');
-                return expect(nearCache.get(ds('key'))).to.equal('val');
+                return nearCache.get(ds('key')).then((res) => {
+                    return expect(res).to.equal('val');
+                });
             });
 
 
             it('returns undefined for non existing value', function () {
-                nearCache.get(ds('random'));
-                return expect(nearCache.getStatistics().missCount).to.equal(1);
+                return nearCache.get(ds('random')).then(() => {
+                    return expect(nearCache.getStatistics().missCount).to.equal(1);
+                });
             });
 
             it('record does not expire if ttl is 0', function () {

